controller/controller.go

```diff
@@ -43,98 +47,68 @@ func New(producer producers.Producer, consumer consumers.Consumer, options *Opti
 		producer: producer,
 		consumer: consumer,
 		options:  options,
-	}
-}
-
-func (c *Controller) Synchronize() error {
-	c.wg.Add(1)
-
-	go func() {
-		for {
-			log.Infoln("Synchronizing DNS entries...")
-
-			endpoints, err := c.producer.Endpoints()
-			if err != nil {
-				log.Fatalf("Error getting endpoints from producer: %v", err)
-			}
 
-			err = c.consumer.Sync(endpoints)
-			if err != nil {
-				log.Fatalf("Error consuming endpoints: %v", err)
-			}
-
-			time.Sleep(c.options.syncPeriod)
-		}
-	}()
-
-	return nil
-}
-
-func (c *Controller) Watch() error {
-	channel, err := c.monitorProducer()
-	if err != nil {
-		return err
+		results: make(chan *pkg.Endpoint),
+		errors:  make(chan error),
+		done:    make(chan struct{}),
 	}
+}
 
-	err = c.processUpdates(channel)
-	if err != nil {
-		return err
-	}
+func (c *Controller) Run() chan error {
+	go c.Synchronize()
+	go c.Watch()
 
-	return nil
+	return c.errors
 }
 
 func (c *Controller) Wait() {
+	signalChan := make(chan os.Signal, 1)
+	signal.Notify(signalChan, syscall.SIGINT, syscall.SIGTERM)
+
+	<-signalChan
+	log.Info("Shutdown signal received, exiting...")
+	close(c.done)
 	c.wg.Wait()
 }
 
-func (c *Controller) monitorProducer() (chan *pkg.Endpoint, error) {
+func (c *Controller) Synchronize() {
 	c.wg.Add(1)
+	defer c.wg.Done()
+
+	for {
+		log.Debugf("[Synchronize] Sleeping for %s...", c.options.syncPeriod)
+		select {
+		case <-time.After(c.options.syncPeriod):
+		case <-c.done:
+			log.Info("[Synchronize] Exited synchronization loop.")
+			return
+		}
 
-	go func() {
-		defer c.wg.Done()
-
-		for {
-			err := c.producer.StartWatch()
-			switch {
-			case err == pkg.ErrEventChannelClosed:
-				log.Debugln("Unable to read from channel. Channel was closed. Trying to restart watch...")
-			case err != nil:
-				log.Fatalln(err)
-			}
+		log.Infoln("[Synchronize] Synchronizing DNS entries...")
+
+		endpoints, err := c.producer.Endpoints()
+		if err != nil {
+			c.errors <- fmt.Errorf("[Synchronize] Error getting endpoints from producer: %v", err)
+			continue
 		}
-	}()
 
-	channel, err := c.producer.ResultChan()
-	if err != nil {
-		return nil, err
+		err = c.consumer.Sync(endpoints)
+		if err != nil {
+			c.errors <- fmt.Errorf("[Synchronize] Error consuming endpoints: %v", err)
+			continue
+		}
 	}
-
-	return channel, nil
 }
 
-func (c *Controller) processUpdates(channel chan *pkg.Endpoint) error {
-	c.wg.Add(1)
-
-	go func() {
-		defer c.wg.Done()
-
-		log.Infoln("Listening for events...")
-
-		for {
-			endpoint, ok := <-channel
-			if !ok {
-				log.Fatalln(errors.New("channel closed"))
-			}
-
-			log.Infof("Processing (%s, %s, %s)\n", endpoint.DNSName, endpoint.IP, endpoint.Hostname)
+func (c *Controller) Watch() {
+	go c.monitorProducer()
```

Comment:
I think I had this implementation before and we settled on passing the channels to each function respectively (further below)